odl/solvers/functional/derivatives.py

```diff
@@ -0,0 +1,326 @@
+        else:
+            self.step = float(step)
+
+        self.method = str(method).lower()
+        if method not in ('backward', 'forward', 'central'):
+            raise ValueError('`method` not understood')
+
+        Operator.__init__(self, functional.domain, functional.domain,
+                          linear=functional.is_linear)
+
+    def _call(self, x):
+        """Return ``self(x)``."""
+        # The algorithm takes finite differences in one dimension at a time
+        # reusing the dx vector to improve efficiency.
+        dfdx = self.domain.zero()
+        dx = self.domain.zero()
+
+        if self.method == 'backward':
+            fx = self.functional(x)
+            for i in range(self.domain.size):
+                dx[i - 1] = 0
+                dx[i] = self.step
+                dfdx[i] = fx - self.functional(x - dx)
+        elif self.method == 'forward':
+            fx = self.functional(x)
+            for i in range(self.domain.size):
+                dx[i - 1] = 0  # reset step from last iteration
+                dx[i] = self.step
+                dfdx[i] = self.functional(x + dx) - fx
+        elif self.method == 'central':
+            for i in range(self.domain.size):
+                dx[i - 1] = 0  # reset step from last iteration
+                dx[i] = self.step / 2
+                dfdx[i] = self.functional(x + dx) - self.functional(x - dx)
+        else:
+            raise RuntimeError('unknown method')
+
+        dfdx /= self.step
+        return dfdx
+
+    def derivative(self, point):
+        """Return the derivative in point.
+
+        The derivative of the gradient is often called the hessian.
+
+        Parameters
+        ----------
+        point : `domain` `element-like`
+            The point that the derivative should be taken in.
+
+        Returns
+        -------
+        derivative : `NumericalDerivative`
+            Numerical estimate of the derivative. Uses the same method as this
+            operator does, but with half the number of significant digits in
+            the step size in order to preserve numerical stability.
+
+        Examples
+        --------
+        Compute a numerical estimate of the derivative of the L2^2 norm:
+
+        >>> import odl
+        >>> space = odl.rn(3)
+        >>> func = odl.solvers.L2NormSquared(space)
+        >>> grad = NumericalGradient(func)
+        >>> hess = grad.derivative([1, 1, 1])
+        >>> hess([1, 0, 0])
+        rn(3).element([2.0, 0.0, 0.0])
+
+        Find the hessian matrix:
+
+        >>> odl.matrix_representation(hess)
+        array([[ 2.,  0.,  0.],
+               [ 0.,  2.,  0.],
+               [ 0.,  0.,  2.]])
+        """
+        return NumericalDerivative(self, point,
+                                   method=self.method, step=np.sqrt(self.step))
+
+if __name__ == '__main__':
+    # pylint: disable=wrong-import-position
+    from odl.util.testutils import run_doctests
+    run_doctests()
```

Comment:
The doctests are all good, but some tests with more complex examples and especially data types (including complex!) feel necessary.